### DISCUSSION TITLE
What's the difference with other license header checker tools?

### DISCUSSION THREAD

**Discussion by xxchan** [korandoru/hawkeye#31] [Q&A] [answered]
e.g., SkyWalking Eyes. What's the goals and non-goals of hawkeye? 👀

**Comment by tisonkun** [comment, marked as answer]
@xxchan Thanks for your question!

First of all, I'd clarify that this tool doesn't want to become the _best_ or _de-facto_ choice in the domain. Thus, I share the background and make some comments primarily for user understanding instead of comparison.

Just days ago, I encountered the impact of https://github.com/apache/skywalking/issues/10223.

I'm [a contributor for SkyWalking Eyes](https://github.com/apache/skywalking-eyes/pulls/tisonkun). But recently, I learned [license-maven-plugin](https://github.com/mathieucarbou/license-maven-plugin), which Apache Pulsar uses. And it doesn't have the issue I met by using SkyWalking Eyes. So fairly, I switched my Java projects to that plugin and rescue.

As you see, the `license-maven-plugin` is a Maven plugin and cannot be used beyond that ecosystem. I recently learned to build a native CLI with picocli and GraalVM's Native Image, so now I know everything to bring license-maven-plugin to native world. And I just do it and you now find HawkEye.

Also, both SkyWalking Eyes and the license-maven-plugin provide the functionality to check dependencies licenses, while I believe it's a totally different topic and should be delivered standalone. By focusing on license header checks I get exact everything I need in HawkEye and it now serves several projects under this org as expected.

So, the major reason for this project is that compared to patching SkyWalking Eyes, my current passion leads me to build a new solution in the same two days while examining my understanding of the Java ecosystem and bringing it to the native world.

Note that the native image of HawkEye is ~20MB while SkyWalking Eyes is also ~25MB, and the Docker Image is ~90MB (with #17 it can be ~20MB) while SkyWalking Eyes is > 500MB. Again, license-maven-plugin can be used only in the Maven plugin ecosystem.

**Reply by tisonkun** [reply]
@xxchan FWIW, the interesting parts I found in this project include the following:

* Build a command-line application with [picocli](https://picocli.info/). My personal experience is that picocli is the best solution.
* Build a native Java application with [GraalVM's Native Image](https://www.graalvm.org/latest/reference-manual/native-image/). While SkyWalking Eyes is written in Golang and can provide cross-platform executables, HawkEye, as a Java native program, achieves the goal with Docker. Refs: #17 
* Build a tool for developers, especially thinking of dev experience and how to expose configs properly. Refs: #22 #32
* Playing with config override and path matching, especially experiencing how gitignore and other famous config conventions do. Refs: #20 

For me, developer experience is a crucial focus and what I can deliver to the community. And providing tools and integrations for where/how developers do their daily work is an essential part.
